packages/interactivity-router/src/index.ts

```diff
@@ -21,23 +21,50 @@ const {
 	'I acknowledge that using private APIs means my theme or plugin will inevitably break in the next version of WordPress.'
 );
 
+interface NavigateOptions {
+	force?: boolean;
+	html?: string;
+	replace?: boolean;
+	timeout?: number;
+	loadingAnimation?: boolean;
+	screenReaderAnnouncement?: boolean;
+}
+
+interface PrefetchOptions {
+	force?: boolean;
+	html?: string;
+}
+
+interface VdomParams {
+	vdom?: typeof initialVdom;
+}
+
+interface Page {
+	regions: Record< string, any >;
+	head: HTMLElement[];
+	title: string;
+	initialData: any;
+}
+
+type RegionsToVdom = ( dom: Document, params?: VdomParams ) => Promise< Page >;
+
 // Check if the navigation mode is full page or region based.
-const navigationMode =
+const navigationMode: 'regionBased' | 'fullPage' =
 	getConfig( 'core/router' ).navigationMode ?? 'regionBased';
 
 // The cache of visited and prefetched pages, stylesheets and scripts.
-const pages = new Map();
+const pages = new Map< string, Promise< Page | false > >();
 const headElements = new Map();
 
 // Helper to remove domain and hash from the URL. We are only interesting in
 // caching the path and the query.
-const getPagePath = ( url ) => {
-	const u = new URL( url, window.location );
+const getPagePath = ( url: string ) => {
+	const u = new URL( url, window.location.href );
```

Comment:
The `URL()` constructor automatically converts `window.location` to string using `window.location.toString()`. However, TS is not happy about it so I explicitly pass `href`.